### ISSUE TITLE
Revamp things-advanced (WSGI/ASGI) examples

### ISSUE THREAD

**Issue by vytas7** [falconry/falcon#2309]
We ought to showcase media handling in these instead of manually wrangling (de-)serialization. It is fine to show what is possible, but the emphasis should still be on media.

Revise App vs API usage: `Useful for debugging problems in your API;` :arrow_right: `Useful for debugging problems in your App;`.

Also check what's going on with the missing `Content-Length` in the ASGI variant, it looks like the proposed middleware doesn't handle it particularly well.

**Comment by abdalahsalah** [comment]
can you assign me for this issue and tell me what is the related part of the repo can i change in @vytas7 
and thank you !

**Comment by vytas7** [comment]
Hi @abdalahsalah!
We don't really "assign" in any special way, your comment here is normally enough for us and others to know you are working on this issue. I can remove the relevant label though.

You need to change `README.rst` and `examples/things_advanced(_asgi)?.py`. it would be nice to deduplicate or at least add a CI check they are in sync, but it is outside of scope for this issue.

**Comment by abdalahsalah** [comment]
Hi @vytas7 ,

Thank you for clarifying how assignments work here. I’ll go ahead and start working on this new issue.

Regarding the two files you mentioned (README.rst and examples/things_advanced(_asgi)?.py), could you please provide more details on the changes expected? Specifically.

**Comment by abdalahsalah** [comment]
@vytas7 ?

**Comment by vytas7** [comment]
Hi again, it is not two files, but three files: [`README.rst`](https://github.com/falconry/falcon/blob/master/README.rst), [`examples/things_advanced.py`](https://github.com/falconry/falcon/blob/master/examples/things_advanced.py), and [`examples/things_advanced_asgi.py`](https://github.com/falconry/falcon/blob/master/examples/things_advanced_asgi.py).

I believe the issue's description explains what to do, I'm not sure if I can be more specific without actually making the changes. Maybe you could start running these files yourself to get familiar with what they do as a starting point?
To summarize the description, we want to better showcase standard [media handling](https://falcon.readthedocs.io/en/stable/api/media.html) out of the box (see also `req.get_media() and `resp.media = ...`).

**Comment by 0xMattB** [comment]
I can handle this one, if you still want these changes implemented.

> We ought to showcase media handling in these instead of manually wrangling (de-)serialization. It is fine to show what is possible, but the emphasis should still be on media.

If I understand, we basically want to completely remove the `JSONTranslator` middleware and change any lines that use `context` to `media`. I played around with these changes in the `things_advanced_asgi` example, and manual testing showed promising results:
- `GET` request worked
- Valid `POST` request worked
- `POST` request with malformed JSON body returned the appropriate error message
- `POST` request with no body returned the appropriate error message

> Also check what's going on with the missing Content-Length in the ASGI variant, it looks like the proposed middleware doesn't handle it particularly well.

I believe this is in reference to a similar problem I fixed in Issue #678, in `process_request()`. From some manual testing, a `GET` requests fails because `req.content_length` is `None` instead of `0`.

```
# examples/things_advanced_asgi.py
# JSONTranslator.process_request()

# Original Code
if req.content_length == 0:
    return

# Replacement Option 1
if req.content_length in (0, None):
    return

# Replacement Option 2
if req.content_length == 0 or req.method in ('GET'):
    return
```

However, if my understanding of the first issue is correct, then this code would be deleted anyway, so this would be moot.

**Comment by vytas7** [comment]
That sounds good @0xMattB, go ahead!

This issue is open ended, we should somehow strike a balance between making the example short and concise, and showcasing what is possible.

Checking for the maximum content length in the case of ASGI is unfortunately tricky. We should probably add better affordances for that in the framework, so for now it is enough things don't break.

**Comment by 0xMattB** [comment]
@vytas7 As both `req.bounded_stream.read()` and `req.get_media()` consume the stream, these two cannot be used in conjunction (if my understanding is correct).

However, we can simplify the example for `ThingsResource.on_post()` *and* show how to use `req.get_media()` by eliminating `JSONTranslator.process_request()`.

And by keeping `JSONTranslator.process_response()`, we can still showcase what the framework can do via `context` on a `GET` request.

Thoughts?

**Comment by vytas7** [comment]
@0xMattB that sounds good. `JSONTranslator` we can get rid of altogether, but yes, we can have some kind of middleware manipulating media, or passing other kind of information via `req.context` (or `resp.context`).

@CaselIT what do you think?

**Comment by CaselIT** [comment]
we could use a middleware to:
- add a request id via process_request
- add some server side timing https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Server-Timing by say timing how long it takes between the 3 middleware methods and then render a total, the routing time and processing time?

**Comment by 0xMattB** [comment]
I wrote something where the middleware captures the `req.path` value via `context`, and uses that to generate a `POST` response body (as the current version of the example does not provide a response body):

```
{'written-to': '/1/things'}
```

Would something like that capture the essence and simplicity of what we're trying to show?

**Comment by Chaz0rk** [comment]
Hi! I’d like to work on this if it’s still available. I’ve set up the development environment and reproduced the issue.

**Comment by vytas7** [comment]
> Hi! I’d like to work on this if it’s still available. I’ve set up the development environment and reproduced the issue.

Sure, go ahread @Chaz0rk! I'm not sure what you mean by "reproduced the issue"... it is not a bug that needs a reproducer or anything like that 😕

In addition, since this is a "good first issue", try not to use [LLM](https://falcon.readthedocs.io/en/latest/community/contributing.html#use-of-llms-ai)s at all in the first iteration.

**Comment by Chaz0rk** [comment]
Hi @vytas7 , thanks for the quick clarification!
I said "reproduced the issue" because when I ran the ASGI example with a simulated GET without a Content-Lenght header I got a `400 Bad Request: Empty request body`.  
But I've been Inaccurate, I did not mean that's bug requiring a reproducer.

I'll follow your guidance and I won't use any LLM to implement the required change.  :slightly_smiling_face:


> > Hi! I’d like to work on this if it’s still available. I’ve set up the development environment and reproduced the issue.
> 
> Sure, go ahread [@Chaz0rk](https://github.com/Chaz0rk)! I'm not sure what you mean by "reproduced the issue"... it is not a bug that needs a reproducer or anything like that 😕
> 
> In addition, since this is a "good first issue", try not to use [LLM](https://falcon.readthedocs.io/en/latest/community/contributing.html#use-of-llms-ai)s at all in the first iteration.

**Comment by vytas7** [comment]
> <...> ASGI example with a simulated GET without a Content-Lenght header I got a `400 Bad Request: Empty request body`.

Oh, you're right, that's bordering a bug indeed! I forgot about that quirk.
